testsuite/forge/src/backend/k8s/swarm.rs

```diff
@@ -286,63 +285,106 @@ impl TryFrom<Service> for KubeService {
+async fn list_stateful_sets(client: K8sClient, kube_namespace: &str) -> Result<Vec<StatefulSet>> {
+    let stateful_set_api: Api<StatefulSet> = Api::namespaced(client, kube_namespace);
     let lp = ListParams::default();
-    let services = node_api.list(&lp).await?.items;
-    services.into_iter().map(KubeService::try_from).collect()
+    let stateful_sets = stateful_set_api.list(&lp).await?.items;
+    Ok(stateful_sets)
 }
 
-fn get_free_port() -> u32 {
-    // get a free port
-    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
-    listener.local_addr().unwrap().port() as u32
+fn stateful_set_name_matches(sts: &StatefulSet, suffix: &str) -> bool {
+    if let Some(s) = sts.metadata.name.as_ref() {
+        s.contains(suffix)
+    } else {
+        false
+    }
+}
+
+fn parse_service_name_from_stateful_set_name(
+    stateful_set_name: &str,
+    enable_haproxy: bool,
+) -> String {
+    let re = Regex::new(r"(aptos-node-\d+)-(validator|fullnode)").unwrap();
+    let cap = re.captures(stateful_set_name).unwrap();
+    let service_base_name = format!("{}-{}", &cap[1], &cap[2]);
+    if enable_haproxy {
+        format!("{}-{}", &service_base_name, HAPROXY_SERVICE_SUFFIX)
+    } else {
+        service_base_name
+    }
+}
+
+fn get_k8s_node_from_stateful_set(
+    sts: &StatefulSet,
+    enable_haproxy: bool,
+    use_port_forward: bool,
+) -> K8sNode {
+    let stateful_set_name = sts.metadata.name.as_ref().unwrap();
+    // If HAProxy is enabled, use its Service name. Otherwise the Service name matches the StatefulSet name
+    let service_name = parse_service_name_from_stateful_set_name(stateful_set_name, enable_haproxy);
+
+    // the full service name includes the namespace
+    let namespace = sts.metadata.namespace.as_ref().unwrap();
+    let full_service_name = format!("{}.{}.svc", &service_name, &namespace);
+
+    // If HAProxy is enabled, use the port on its Service. Otherwise use the port on the validator Service
+    let mut rest_api_port = if enable_haproxy {
+        REST_API_HAPROXY_SERVICE_PORT
+    } else {
+        REST_API_SERVICE_PORT
+    };
+
+    if use_port_forward {
+        rest_api_port = get_free_port();
+    }
+    let index = parse_node_index(stateful_set_name).expect("error to parse node index");
+
+    // Extract the image tag from the StatefulSet spec
+    let image_tag = sts
+        .spec
+        .as_ref()
+        .unwrap()
+        .template
+        .spec
+        .as_ref()
+        .unwrap()
+        .containers[0]
+        .image
+        .as_ref()
+        .unwrap()
+        .split(':')
+        .collect::<Vec<&str>>()[1];
+
+    // Extract the namespace from the StatefulSet spec
+    let kube_namespace = sts.metadata.namespace.as_ref().unwrap();
+
+    K8sNode {
+        name: stateful_set_name.clone(),
+        stateful_set_name: stateful_set_name.clone(),
+        // TODO: fetch this from running node
+        peer_id: PeerId::random(),
```

Comment:
Where is this being used?

Reply:
It's being used in a bunch of local tests and also to target operations on a single node. We've just always used a random peer_id for k8s nodes though we should really figure out how to use the real one.